config/src/main/java/com/typesafe/config/ConfigReferenceResolveRequest.java

```diff
@@ -0,0 +1,24 @@
+package com.typesafe.config;
+
+/**
+ * A request for a resolver to return a value for a particular reference.
+ */
+public interface ConfigReferenceResolveRequest {
+
+    /**
+     * The reference's path.
+     */
+    public Path getPath();
+
+    /**
+     * Is this an optional reference?
+     */
+    public boolean isOptional();
+
+    /**
+     * Returns the reference value we're being asked to resolve. Returning this
+     * value will cause the reference to be left unresolved in the config.
+     */
+    public ConfigValue getReferenceValue();
+
```

Comment:
Does the resolver need to know about optional - it could return null for not found, and let the library throw an exception if not optional?

I feel like this interface could be replaced with parameters to the lookup method in ConfigReferenceResolver?
`ConfigValue lookup(String path)` ? 
If we needed other parameters in the future we could add a variant of `ConfigReferenceResolver` perhaps.

Reply:
The reason I did it this way was mainly so that the reference value was available so there was a way for resolvers to express leaving the references unresolved. But I guess you can achieve a more coarse-grained version of the same thing by setting allowUnresolved=true and returning null.